compiler/rustc_mir_transform/src/fold.rs

```diff
@@ -0,0 +1,360 @@
+        Some(StatementKind::Assign(Box::new((*final_place, Rvalue::Use(Operand::Move(new_place))))))
+    } else {
+        Some(StatementKind::Assign(Box::new((*final_place, Rvalue::Use(Operand::Copy(new_place))))))
+    }
+}
+
+// Fold:
+// _2 = _1 as *const T;
+// _3 = _2 as *const U;
+// Into:
+// _3 = _1 as *const U;
+fn fold_ptr_ptr_cast<'tcx>(
+    _tcx: TyCtxt<'tcx>,
+    temp_place: &Place<'tcx>,
+    temp_rvalue: &Rvalue<'tcx>,
+    final_place: &Place<'tcx>,
+    final_rvalue: &Rvalue<'tcx>,
+) -> Option<StatementKind<'tcx>> {
+    if !temp_place.projection.is_empty() {
+        return None;
+    }
+    let Rvalue::Cast(CastKind::PtrToPtr | CastKind::Pointer(PointerCast::MutToConstPointer), temp_operand, _temp_ty) = temp_rvalue else { return None; };
+    let Rvalue::Cast(CastKind::PtrToPtr | CastKind::Pointer(PointerCast::MutToConstPointer), _final_operand, final_ty) = final_rvalue else { return None; };
+
+    Some(StatementKind::Assign(Box::new((
+        *final_place,
+        Rvalue::Cast(CastKind::PtrToPtr, temp_operand.clone(), *final_ty),
+    ))))
+}
+
+fn did_optimization<'tcx>(body: &mut Body<'tcx>, tcx: TyCtxt<'tcx>) -> bool {
+    let mut visitor = AnalysisVisitor {
+        analysis: IndexVec::from_elem_n(Analysis::default(), body.local_decls.len()),
+    };
+    visitor.visit_body(body);
+    let analysis = visitor.analysis;
+
+    let mut invalidated_statements = FxHashSet::default();
+
+    'outer: for result in analysis.iter() {
+        let write = result.write.as_assign(body);
+        let read = result.read.as_assign(body);
+        let Some(((temp_loc, (temp_place, temp_rvalue)), (final_loc, (final_place, final_rvalue)))) = write.zip(read) else {
+                continue;
+            };
+        // We only apply this optimization within a block, this means that we get to skip all
+        // reasoning about flow control in our analysis.
+        if temp_loc.block != final_loc.block {
+            continue;
+        }
+
+        // If this is a single statement, it's an assignment that modifies a variable based on
+        // itself, such as an AddAssign impl.
+        if temp_loc == final_loc {
+            continue;
+        }
+
+        // If the creation of the temporary comes after the use of it, then by whatever means this
+        // is not the pattern we are looking for.
+        if temp_loc.statement_index > final_loc.statement_index {
+            continue;
+        }
+
+        if invalidated_statements.contains(&temp_loc) || invalidated_statements.contains(&final_loc)
+        {
+            continue;
+        }
+
+        // We only do optimizations where there are no statements other than storage markers
+        // between the two statements. This is our proxy for alias analysis: if there are no
+        // statements between the creation of the temporary and its use, nothing can
+        // observe us eliding the temporary.
```

Comment:
Sadly we don't really have a good alternative, but I wish we did.  This restriction keeps this change from optimizing the <https://rust.godbolt.org/z/5W5cno4EW> pattern that shows up pretty often.

Reply:
Maybe this is too wild, but you could insert new locals which read from the references, replace all accesses through the references with those locals, and go from there. I've considered doing this, but I think I have higher-yield things on my plate right now and I also have no idea how to prevent this from becoming a de-optimization.